### PR TITLE
[LI-HOTFIX] Adding new metrics for MetadataAllTopics request rate and response size distribution

### PR DESCRIPTION
**Pull request by gitlw**
TICKET = LIKAFKA-39573
LI_DESCRIPTION =
Currently, Kafka emits a metric to show the rate of Metadata requests.

However, not all metadata requests are equal. If the topics field of the Metadata request is null, it indicates that the Metadata request is retrieving the metadata of all topics. In this PR, we add a new metric to indicate the rate of such topics=null Metadata requests.

Besides, we are adding another metric to show the size distribution of Metadata responses sent back to clients. This will help us identify brokers that are overloaded with large Metadata responses.

EXIT_CRITERIA = When this change is merged in upstream Kafka and pulled into this repo.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 1]
```diff
@@ -3048,6 +3048,7 @@ class KafkaApis(val requestChannel: RequestChannel,
     val response = responseOpt match {
       case Some(response) =>
         val responseSend = request.context.buildResponse(response)
+        request.responseBytes = responseSend.size()
```
Comment (groelofs) [inline core/src/main/scala/kafka/server/KafkaApis.scala:3051]: Just checking since I can't easily tell here:  does size() return bytes rather than some object count?  (Would make sense if the response is byte[] ...)

File: core/src/main/scala/kafka/network/RequestChannel.scala [hunk 2]
```diff
@@ -483,6 +494,7 @@ class RequestMetrics(name: String) extends KafkaMetricsGroup {
     removeMetric(TotalTimeMs, tags)
     removeMetric(ResponseSendTimeMs, tags)
     removeMetric(RequestBytes, tags)
+    removeMetric(ResponseBytes, tags)
```
Comment (groelofs) [inline core/src/main/scala/kafka/network/RequestChannel.scala:497]: Doesn't MetadataAllTopics need to be removed, too?  Or is this related to the histogram part?

**Review by groelofs — COMMENTED** [review]
I don't quite understand how the counter works, but it uses the same pattern as a previous one, so I'll chalk it up as something to dig into more later.

Just a couple of questions, probably also reflecting my ignorance of some subtleties...

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 1 above)
Comment (gitlw) [reply core/src/main/scala/kafka/server/KafkaApis.scala:3051]: The responseSend will be an object of type NetworkSent, which is a subclass of ByteBufferSend.
The ByteBufferSend.size() is returning the size of the serialized buffer.
https://github.com/linkedin/kafka/blob/d0e436c471ba4122ddcc0f7a1624546f97c4a517/clients/src/main/java/org/apache/kafka/common/network/ByteBufferSend.java#L55

File: core/src/main/scala/kafka/network/RequestChannel.scala (hunk 2 above)
Comment (gitlw) [reply core/src/main/scala/kafka/network/RequestChannel.scala:497]: The full metric names are like
kafka.network:type=RequestMetrics,name=RequestBytes,request=MetadataAllTopics
kafka.network:type=RequestMetrics,name=ResponseBytes,request=MetadataAllTopics

So there would be one RequestMetrics object with a tags field being Map("request" -> RequestMetrics.MetadataAllTopics)
and inside that object we have the list of fields including
RequestBytes and ResponseBytes etc. After removing these individual fields, every metric with the tag "request=MetadataAllTopics" will be gone.

Comment (groelofs) [reply core/src/main/scala/kafka/network/RequestChannel.scala:497]: Ahhh, I see:  these are effectively different dimensions, and the rate was already present.

Since it appears that ResponseBytes gets added to all of the categories/tags, not just the MetadataAllTopics one, is there any concern about overall metrics sizes/rates?  (In particular, are the other tags much broader/more common than the all-topics metadata one?)

**Review by groelofs — APPROVED** [review]
LGTM

File: core/src/main/scala/kafka/network/RequestChannel.scala [hunk 3]
```diff
@@ -155,7 +157,11 @@ object RequestChannel extends Logging {
           )
         }
         else Seq.empty
-      val metricNames = fetchMetricNames :+ header.apiKey.name
+      val metadataMetricNames =
```
Comment (lmr3796) [inline core/src/main/scala/kafka/network/RequestChannel.scala:160]: (Feel free to drop) nit: The approach to ensure only metadata response size is updated here may not be straightforward for newcomers to the broker (using `Seq.empty` w/ `forEach`).  Maybe add a one-line comment for that?
